File: src/engine/analyze.ts
import pack from "../../rules/rulepack.json";
import type {
  AgentNote,
  AnalysisInput,
  CheckFn,
  Finding,
  Grade,
  Report,
  RuleMeta,
  RulePack,
  RulePenalty,
} from "./types";
import { buildContext, estimateTokens } from "./parse";
import {
  boundariesMissing,
  codexCap,
  contradictionLite,
  directoryTree,
  duplicateRules,
  emphasisOveruse,
  enforcementProse,
  lintLeakage,
  noCommands,
  pointerWithoutInstruction,
  secretLeak,
  sizeCost,
  vagueRules,
} from "./checks/file-checks";
import {
  bridgeMissing,
  deadCommand,
  importResolution,
  readmeRedundancy,
  stalePath,
} from "./checks/repo-checks";

export const rulepack = pack as RulePack;

const IMPLS: Record<string, CheckFn> = {
  "lint-leakage": lintLeakage,
  "directory-tree": directoryTree,
  "vague-rules": vagueRules,
  "no-commands": noCommands,
  "enforcement-prose": enforcementProse,
  "secret-leak": secretLeak,
  "emphasis-overuse": emphasisOveruse,
  "pointer-without-instruction": pointerWithoutInstruction,
  "size-cost": sizeCost,
  "duplicate-rules": duplicateRules,
  "contradiction-lite": contradictionLite,
  "boundaries-missing": boundariesMissing,
  "codex-cap": codexCap,
  "dead-command": deadCommand,
  "stale-path": stalePath,
  "readme-redundancy": readmeRedundancy,
  "bridge-missing": bridgeMissing,
  "import-resolution": importResolution,
};

/**
 * Rulepack and implementations must agree exactly — a rule with no code and
 * code with no rule are both configuration drift, and drift throws.
 */
export function validatePack(rp: RulePack, impls: Record<string, CheckFn>): void {
  const packIds = new Set(rp.rules.map((r) => r.id));
  const implIds = new Set(Object.keys(impls));
  for (const id of packIds) {
    if (!implIds.has(id)) throw new Error(`rulepack rule "${id}" has no implementation`);
  }
  for (const id of implIds) {
    if (!packIds.has(id)) throw new Error(`implementation "${id}" is not in the rulepack`);
  }
  for (const r of rp.rules) {
    // https-only is a security invariant, not a style rule: evidenceUrl lands
    // in an href, and a javascript:/data: value there would be XSS if a rule
    // pack ever sourced URLs from outside the bundle.
    if (!/^https:\/\//.test(r.evidenceUrl))
      throw new Error(`rule "${r.id}" evidenceUrl must be https://`);
    if (r.weight <= 0 || r.maxPenalty <= 0)
      throw new Error(`rule "${r.id}" has non-positive weight/maxPenalty`);
  }
}

validatePack(rulepack, IMPLS);

function gradeOf(score: number): Grade {
  if (score >= 90) return "A";
  if (score >= 75) return "B";
  if (score >= 60) return "C";
  if (score >= 40) return "D";
  return "F";
}

function agentNotes(input: AnalysisInput, lineCount: number, byteSize: number): AgentNote[] {
  const notes: AgentNote[] = [];
  const isClaude = /claude\.md$/i.test(input.fileName);
  const isAgents = /agents\.md$/i.test(input.fileName);
  notes.push({
    agent: "Claude Code",
    note:
      (isAgents
        ? "Does not read AGENTS.md natively — needs a CLAUDE.md containing `@AGENTS.md`. "
        : "") +
      `Injects ${isClaude ? "this file" : "CLAUDE.md"} as an advisory user message (no compliance guarantee); official size target is under 200 lines${lineCount > 200 ? ` — this file is ${lineCount}` : ""}. Zero-exception rules belong in hooks.`,
  });
  notes.push({
    agent: "OpenAI Codex",
    note:
      byteSize > 32 * 1024
        ? `Truncates at 32 KiB — this file is ${(byteSize / 1024).toFixed(1)} KiB, everything past the cap is silently dropped.`
        : `Reads AGENTS.md natively (nearest file wins, 32 KiB cap — this file is ${(byteSize / 1024).toFixed(1)} KiB). Supports AGENTS.override.md for local tweaks.`,
  });
  notes.push({
    agent: "Cursor",
    note: "Reads AGENTS.md natively. Path-scoped rules need .cursor/rules/*.mdc — plain .md files there are silently ignored.",
  });
  notes.push({
    agent: "GitHub Copilot",
    note: "Reads AGENTS.md natively (coding agent, since Aug 2025) alongside .github/copilot-instructions.md.",
  });
  return notes;
}

export function analyze(input: AnalysisInput): Report {
  const ctx = buildContext(input.fileName, input.content, input.repo);
  const mode = input.repo ? "repo" : "file";

  const findings: Finding[] = [];
  const penalties: RulePenalty[] = [];
  let totalPenalty = 0;

  // Escape hatch for heuristics: an `aispekt-ignore` marker (legacy: prune-ignore) on the finding's
  // line or the line above suppresses it (like eslint-disable-next-line).
  const isIgnored = (line: number): boolean =>
    /(aispekt|prune)-ignore/.test(ctx.lines[line - 1] ?? "") ||
    /(aispekt|prune)-ignore/.test(ctx.lines[line - 2] ?? "");

  for (const rule of rulepack.rules) {
    if (rule.scope === "repo" && !input.repo) continue;
    const impl = IMPLS[rule.id];
    if (!impl) throw new Error(`no implementation for rule "${rule.id}"`);
    const raw = impl(ctx).filter((f) => !isIgnored(f.line));
    if (raw.length === 0) continue;
    for (const f of raw) {
      findings.push({
        ...f,
        ruleId: rule.id,
        ruleName: rule.name,
        severity: rule.severity,
        recommendation: f.recommendation ?? rule.recommendation,
        evidenceUrl: rule.evidenceUrl,
        evidenceTier: rule.evidenceTier,
      });
    }
    const penalty = Math.min(rule.maxPenalty, raw.length * rule.weight);
    penalties.push({ ruleId: rule.id, count: raw.length, penalty });
    totalPenalty += penalty;
  }

  const score = Math.max(0, Math.min(100, Math.round(100 - totalPenalty)));
  const byteSize = new TextEncoder().encode(input.content).length;

  const severityOrder = { error: 0, warn: 1, info: 2 } as const;
  findings.sort(
    (a, b) =>
      severityOrder[a.severity] - severityOrder[b.severity] ||
      a.line - b.line ||
      a.ruleId.localeCompare(b.ruleId),
  );

  const notices: string[] = [];
  if (input.repo?.truncated) {
    notices.push(
      `Repo walk hit the ${input.repo.paths.length.toLocaleString()}-path cap — stale-path and @-import drift checks were skipped rather than run against a partial file list.`,
    );
  }

  return {
    fileName: input.fileName,
    mode,
    score,
    grade: gradeOf(score),
    lineCount: ctx.lines.length,
    byteSize,
    tokenEstimate: estimateTokens(input.content),
    findings,
    penalties,
    agentNotes: agentNotes(input, ctx.lines.length, byteSize),
    notices,
    rulepackVersion: rulepack.version,
    rulepackUpdated: rulepack.updated,
  };
}

export function ruleMeta(id: string): RuleMeta | undefined {
  return rulepack.rules.find((r) => r.id === id);
}

File: crates/core/src/analyze.rs
use crate::parse::{build_context, estimate_tokens};
use crate::render::{js_round, to_fixed1, to_locale_string};
use crate::rulepack::{impl_for, RULEPACK};
use crate::types::{AgentNote, AnalysisInput, Finding, Report, RulePenalty};
use regex::Regex;
use serde_json::{json, Map, Value};
use std::sync::LazyLock;

fn grade_of(score: i64) -> &'static str {
    if score >= 90 {
        "A"
    } else if score >= 75 {
        "B"
    } else if score >= 60 {
        "C"
    } else if score >= 40 {
        "D"
    } else {
        "F"
    }
}

fn agent_notes(input: &AnalysisInput, line_count: usize, byte_size: usize) -> Vec<AgentNote> {
    let lower = input.file_name.to_lowercase();
    let is_claude = lower.ends_with("claude.md");
    let is_agents = lower.ends_with("agents.md");
    let kib = to_fixed1(byte_size as f64 / 1024.0);
    vec![
        AgentNote {
            agent: "Claude Code".to_string(),
            note: format!(
                "{}Injects {} as an advisory user message (no compliance guarantee); official size target is under 200 lines{}. Zero-exception rules belong in hooks.",
                if is_agents {
                    "Does not read AGENTS.md natively — needs a CLAUDE.md containing `@AGENTS.md`. "
                } else {
                    ""
                },
                if is_claude { "this file" } else { "CLAUDE.md" },
                if line_count > 200 {
                    format!(" — this file is {line_count}")
                } else {
                    String::new()
                }
            ),
        },
        AgentNote {
            agent: "OpenAI Codex".to_string(),
            note: if byte_size > 32 * 1024 {
                format!(
                    "Truncates at 32 KiB — this file is {kib} KiB, everything past the cap is silently dropped."
                )
            } else {
                format!(
                    "Reads AGENTS.md natively (nearest file wins, 32 KiB cap — this file is {kib} KiB). Supports AGENTS.override.md for local tweaks."
                )
            },
        },
        AgentNote {
            agent: "Cursor".to_string(),
            note: "Reads AGENTS.md natively. Path-scoped rules need .cursor/rules/*.mdc — plain .md files there are silently ignored.".to_string(),
        },
        AgentNote {
            agent: "GitHub Copilot".to_string(),
            note: "Reads AGENTS.md natively (coding agent, since Aug 2025) alongside .github/copilot-instructions.md.".to_string(),
        },
    ]
}

static IGNORE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(aispekt|prune)-ignore").unwrap());

pub fn analyze(input: &AnalysisInput) -> Report {
    let pack = &*RULEPACK;
    let ctx = build_context(&input.file_name, &input.content, input.repo.clone());
    let mode = if input.repo.is_some() { "repo" } else { "file" };

    let mut findings: Vec<Finding> = Vec::new();
    let mut penalties: Vec<RulePenalty> = Vec::new();
    let mut total_penalty = 0.0f64;

    // Escape hatch for heuristics: an `aispekt-ignore` marker (legacy:
    // prune-ignore) on the finding's line or the line above suppresses it.
    let is_ignored = |line: usize| -> bool {
        let at = |idx: usize| {
            idx.checked_sub(1)
                .and_then(|i| ctx.lines.get(i))
                .map(|l| IGNORE_RE.is_match(l))
                .unwrap_or(false)
        };
        at(line) || at(line.wrapping_sub(1))
    };

    for rule in &pack.rules {
        if rule.scope == "repo" && input.repo.is_none() {
            continue;
        }
        let check = impl_for(&rule.id)
            .unwrap_or_else(|| panic!("no implementation for rule \"{}\"", rule.id));
        let raw: Vec<_> = check(&ctx)
            .into_iter()
            .filter(|f| !is_ignored(f.line))
            .collect();
        if raw.is_empty() {
            continue;
        }
        let count = raw.len();
        for f in raw {
            findings.push(Finding {
                line: f.line,
                excerpt: f.excerpt,
                message: f.message,
                raw_recommendation: f.recommendation.is_some(),
                rule_id: rule.id.clone(),
                rule_name: rule.name.clone(),
                severity: rule.severity,
                recommendation: f.recommendation.unwrap_or_else(|| rule.recommendation.clone()),
                evidence_url: rule.evidence_url.clone(),
                evidence_tier: rule.evidence_tier,
            });
        }
        let penalty = rule.max_penalty.min(count as f64 * rule.weight);
        penalties.push(RulePenalty {
            rule_id: rule.id.clone(),
            count,
            penalty,
        });
        total_penalty += penalty;
    }

    let score = js_round(100.0 - total_penalty).clamp(0.0, 100.0) as i64;
    let byte_size = input.content.len();

    findings.sort_by(|a, b| {
        a.severity
            .order()
            .cmp(&b.severity.order())
            .then(a.line.cmp(&b.line))
            .then(a.rule_id.cmp(&b.rule_id))
    });

    let mut notices: Vec<String> = Vec::new();
    if let Some(repo) = &input.repo {
        if repo.truncated {
            notices.push(format!(
                "Repo walk hit the {}-path cap — stale-path and @-import drift checks were skipped rather than run against a partial file list.",
                to_locale_string(repo.paths.len())
            ));
        }
    }

    Report {
        file_name: input.file_name.clone(),
        mode,
        score,
        grade: grade_of(score),
        line_count: ctx.lines.len(),
        byte_size,
        token_estimate: estimate_tokens(&input.content),
        findings,
        penalties,
        agent_notes: agent_notes(input, ctx.lines.len(), byte_size),
        notices,
        rulepack_version: pack.version.clone(),
        rulepack_updated: pack.updated.clone(),
    }
}

fn number(v: f64) -> Value {
    if v.fract() == 0.0 && v.abs() < 9.0e15 {
        json!(v as i64)
    } else {
        json!(v)
    }
}

/// Serialize a report exactly as the TS engine's `JSON.stringify(report)`
/// would: same key order, including the spread-position quirk where a raw
/// per-finding recommendation surfaces before ruleId.
pub fn report_to_value(report: &Report) -> Value {
    let findings: Vec<Value> = report
        .findings
        .iter()
        .map(|f| {
            let mut m = Map::new();
            m.insert("line".into(), json!(f.line));
            m.insert("excerpt".into(), json!(f.excerpt));
            m.insert("message".into(), json!(f.message));
            if f.raw_recommendation {
                m.insert("recommendation".into(), json!(f.recommendation));
            }
            m.insert("ruleId".into(), json!(f.rule_id));
            m.insert("ruleName".into(), json!(f.rule_name));
            m.insert("severity".into(), json!(f.severity.as_str()));
            if !f.raw_recommendation {
                m.insert("recommendation".into(), json!(f.recommendation));
            }
            m.insert("evidenceUrl".into(), json!(f.evidence_url));
            m.insert("evidenceTier".into(), json!(f.evidence_tier.as_str()));
            Value::Object(m)
        })
        .collect();
    let penalties: Vec<Value> = report
        .penalties
        .iter()
        .map(|p| {
            json!({
                "ruleId": p.rule_id,
                "count": p.count,
                "penalty": number(p.penalty),
            })
        })
        .collect();
    let agent_notes: Vec<Value> = report
        .agent_notes
        .iter()
        .map(|n| json!({ "agent": n.agent, "note": n.note }))
        .collect();
    json!({
        "fileName": report.file_name,
        "mode": report.mode,
        "score": report.score,
        "grade": report.grade,
        "lineCount": report.line_count,
        "byteSize": report.byte_size,
        "tokenEstimate": report.token_estimate,
        "findings": findings,
        "penalties": penalties,
        "agentNotes": agent_notes,
        "notices": report.notices,
        "rulepackVersion": report.rulepack_version,
        "rulepackUpdated": report.rulepack_updated,
    })
}

/// `JSON.stringify(report, null, 2)` equivalent.
pub fn report_to_json_pretty(report: &Report) -> String {
    serde_json::to_string_pretty(&report_to_value(report)).expect("report serializes")
}

/// Compact `JSON.stringify(report)` equivalent (wasm/browser interchange).
pub fn report_to_json(report: &Report) -> String {
    serde_json::to_string(&report_to_value(report)).expect("report serializes")
}
